Add CRUD for VirtualRouter top level resource and VirtualRouterPeer c…

naveenchekuri opened the pull request:
…hild resource

<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

This feature is to develop new cmdlets in order to support CRUD operations on new resource called virtualRouter under Microsoft.Network and peerings child resource under virtualRouter

Added following cmdlets

New-AzVirtualRouter
Remove-AzVirtualRouter
Get-AzVirtualRouter
Add-AzVirtualRouterPeer
Remove-AzVirtualRouterPeer
Get-AzVirtualRouterPeer
Update-AzVirtualRouterPeer

## Checklist

- [x] I have read the [_Submitting Changes_](../CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [x] the markdown help files have been regenerated using the commands listed [here](../documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

adxsdkps commented:
Can one of the admins verify this patch?

erich-wang commented:
@naveenchekuri, @number213 , which milestone does this PR target to? Probably won't make it for 10/15 release.

number213 commented:
@erich-wang @naveenchekuri yes, it's too late. I've retargeted it to the next release branch

number213 commented:
@naveenchekuri please pull the latest from `network-october`

number213 commented on src/Network/Network/Common/NetworkResourceManagerProfile.cs:
```diff
@@ -89,6 +89,37 @@ public override string ProfileName
+                    Location = source.Location,
+                    Type = source.Type,
+                    ProvisioningState = source.ProvisioningState,
+                    VirtualRouterAsn = (uint) source.VirtualRouterAsn
```
VirtualRouterAsn is nullable in SDK model, right? This could cause an exception during casting

naveenchekuri replied:
VirtualRouterAsn is long in sdk. I dont think long can be null

number213 commented on src/Network/Network/Common/NetworkResourceManagerProfile.cs:
```diff
@@ -89,6 +89,37 @@ public override string ProfileName
+                    router.VirtualRouterIps = source.VirtualRouterIps.ToList<string>();
+                }
+
+                router.Peerings = new List<CNM.PSVirtualRouterPeer>();
```
You sure you don't need to map all peerings? I get that you with poll for them in Get, but there may be other cases where this model is going to be used

naveenchekuri replied:
I would prefer to fetch all peerings for the virtualrouter. I totally see where you are coming from. For now, I don't believe populating peerings here is needed

number213 commented on src/Network/Network/Generated/Models/PSVirtualRouterPeer.cs:
```diff
@@ -0,0 +1,43 @@
+﻿// <auto-generated>
```
Since it's a manually written model, could you please remove parts about generation from the header?

naveenchekuri replied:
Done

number213 commented on src/Network/Network/VirtualRouter/GetAzureRMVirtualRouterCommand.cs:
```diff
@@ -0,0 +1,118 @@
+                    var vVirtualRouterModel = NetworkResourceManagerProfile.Mapper.Map<CNM.PSVirtualRouter>(vVirtualRouter);
+                    vVirtualRouterModel.ResourceGroupName = NetworkBaseCmdlet.GetResourceGroup(vVirtualRouter.Id);
+                    vVirtualRouterModel.Tag = TagsConversionHelper.CreateTagHashtable(vVirtualRouter.Tags);
+                    foreach (var peering in vVirtualRouter.Peerings)
```
Same comment about using `VirtualRouterPeerings.List`

naveenchekuri replied:
Agreed. Good idea to reduce network calls.Done

naveenchekuri commented on src/Network/Network/help/Add-AzVirtualRouterPeer.md:
```diff
@@ -0,0 +1,200 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.Network.dll-Help.xml
+Module Name: Az.Network
+online version:
```
Done

naveenchekuri commented on src/Network/Network/VirtualRouter/GetAzureRMVirtualRouterCommand.cs:
```diff
@@ -0,0 +1,118 @@
+                foreach (var peering in vVirtualRouter.Peerings)
+                {
+                    var resourceInfo = new ResourceIdentifier(peering.Id);
+                    var vVirtualRouterPeering = this.NetworkClient.NetworkManagementClient.VirtualRouterPeerings.Get(ResourceGroupName, RouterName, resourceInfo.ResourceName);
```
Agreed. Good idea to reduce network calls

naveenchekuri commented on src/Network/Network/Generated/Models/PSVirtualRouter.cs:
```diff
@@ -0,0 +1,58 @@
+﻿// <auto-generated>
```
Done

naveenchekuri commented on src/Network/Network/Common/NetworkResourceManagerProfile.cs:
```diff
@@ -89,6 +89,37 @@ public override string ProfileName
+                    {
+                        Id = source.HostedGateway.Id
+                    };
+                }
```
We are not supporting Hosted Subnet for now and do not want to display this to user (This might confuse the user)

naveenchekuri commented on src/Network/Network/Az.Network.psd1:
```diff
@@ -420,7 +420,10 @@ CmdletsToExport = 'Add-AzApplicationGatewayAuthenticationCertificate',
                'Start-AzVirtualNetworkGatewayConnectionPacketCapture', 
-               'Stop-AzVirtualNetworkGatewayConnectionPacketCapture'
+               'Stop-AzVirtualNetworkGatewayConnectionPacketCapture',
+			   'New-AzVirtualRouter', 'Remove-AzVirtualRouter', 'Get-AzVirtualRouter',
```
Done

naveenchekuri commented on src/Network/Network.Test/ScenarioTests/VirtualRouterTests.cs:
```diff
@@ -0,0 +1,44 @@
+
+        [Fact]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        [Trait(Category.Owner, Category.pgtm)]
```
Done

naveenchekuri commented:
@number213 Thanks so much for help with coding and review. I addressed all your comments. I'm closing this PR as I raised new PR with october branch https://github.com/Azure/azure-powershell/pull/10273. Please review that where I had addressed all the comments. Thanks again!
